File: crates/stateset-embedded/src/print_stations.rs
//! Print station operations (paired agents + print job queue).

use stateset_core::{
    CreatePrintStation, EnqueuePrintJob, PairStationResult, PrintJob, PrintJobFilter, PrintJobId,
    PrintStation, PrintStationId, Result,
};
use stateset_db::{Database, DatabaseCapability};
use std::sync::Arc;

/// Print station operations.
pub struct PrintStations {
    db: Arc<dyn Database>,
}

impl std::fmt::Debug for PrintStations {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PrintStations").finish_non_exhaustive()
    }
}

impl PrintStations {
    pub(crate) fn new(db: Arc<dyn Database>) -> Self {
        Self { db }
    }

    /// Whether print stations are supported by the active backend.
    #[must_use]
    pub fn is_supported(&self) -> bool {
        self.db.supports_capability(DatabaseCapability::PrintStations)
    }

    fn ensure(&self) -> Result<()> {
        self.db.ensure_capability(DatabaseCapability::PrintStations)
    }

    /// Pair a new print station, returning the one-time token.
    pub fn pair(&self, input: CreatePrintStation) -> Result<PairStationResult> {
        self.ensure()?;
        self.db.print_stations().pair(input)
    }

    /// List paired stations.
    pub fn list_stations(&self) -> Result<Vec<PrintStation>> {
        self.ensure()?;
        self.db.print_stations().list_stations()
    }

    /// Get a station by ID.
    pub fn get_station(&self, id: PrintStationId) -> Result<Option<PrintStation>> {
        self.ensure()?;
        self.db.print_stations().get_station(id)
    }

    /// Revoke a station.
    pub fn revoke_station(&self, id: PrintStationId) -> Result<PrintStation> {
        self.ensure()?;
        self.db.print_stations().revoke_station(id)
    }

    /// Enqueue a print job to a station.
    pub fn enqueue_job(
        &self,
        station_id: PrintStationId,
        input: EnqueuePrintJob,
    ) -> Result<PrintJob> {
        self.ensure()?;
        self.db.print_stations().enqueue_job(station_id, input)
    }

    /// Pick up the next queued job for a station.
    pub fn next_job(&self, station_id: PrintStationId) -> Result<Option<PrintJob>> {
        self.ensure()?;
        self.db.print_stations().next_job(station_id)
    }

    /// Mark a job printed or failed.
    pub fn complete_job(&self, job_id: PrintJobId, success: bool) -> Result<PrintJob> {
        self.ensure()?;
        self.db.print_stations().complete_job(job_id, success)
    }

    /// List jobs for a station.
    pub fn list_jobs(
        &self,
        station_id: PrintStationId,
        filter: PrintJobFilter,
    ) -> Result<Vec<PrintJob>> {
        self.ensure()?;
        self.db.print_stations().list_jobs(station_id, filter)
    }
}

File: cli/src/tools/print-stations.js
/**
 * Print Station Tools Module
 *
 * MCP tool definitions for paired print stations and their print job queue.
 */

import { z } from 'zod';
import { applyRequired } from '../utils/apply-guard.js';

const withPolicyDomain = (policyDomain, tools) => tools.map((tool) => ({ policyDomain, ...tool }));

export const printStationTools = withPolicyDomain('print-stations', [
  {
    name: 'list_print_stations',
    description: 'List paired print stations.',
    inputSchema: {},
    permission: 'read',
    handler: async ({ commerce }) => {
      const stations = await commerce.printStations.listStations();
      return { success: true, count: stations.length, stations };
    },
  },
  {
    name: 'get_print_station',
    description: 'Get a print station by ID.',
    inputSchema: {
      id: z.string().min(1).describe('Print station ID'),
    },
    permission: 'read',
    handler: async ({ commerce, params }) => {
      const station = await commerce.printStations.getStation(params.id);
      if (!station) {
        return { success: false, error: 'Print station not found' };
      }
      return { success: true, station };
    },
  },
  {
    name: 'pair_print_station',
    description: 'Pair a new print station. Returns a one-time pairing token.',
    inputSchema: {
      name: z.string().min(1).describe('Station name'),
      printers: z.array(z.string().min(1)).optional().describe('Printer names on the station'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Pair print station', params);
      }
      const result = await commerce.printStations.pair({
        name: params.name,
        printers: params.printers,
      });
      return {
        success: true,
        message: 'Print station paired',
        station: result.station,
        token: result.token,
      };
    },
  },
  {
    name: 'revoke_print_station',
    description: 'Revoke a paired print station.',
    inputSchema: {
      id: z.string().min(1).describe('Print station ID'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Revoke print station', params);
      }
      const station = await commerce.printStations.revokeStation(params.id);
      return { success: true, message: 'Print station revoked', station };
    },
  },
  {
    name: 'list_print_jobs',
    description: 'List print jobs for a station.',
    inputSchema: {
      stationId: z.string().min(1).describe('Print station ID'),
      status: z
        .enum(['queued', 'picked_up', 'printed', 'failed'])
        .optional()
        .describe('Filter by job status'),
      limit: z.number().int().positive().optional().describe('Maximum results'),
      offset: z.number().int().min(0).optional().describe('Results to skip'),
    },
    permission: 'read',
    handler: async ({ commerce, params }) => {
      const jobs = await commerce.printStations.listJobs(params.stationId, {
        status: params.status,
        limit: params.limit,
        offset: params.offset,
      });
      return { success: true, count: jobs.length, jobs };
    },
  },
  {
    name: 'enqueue_print_job',
    description: 'Enqueue a print job to a station.',
    inputSchema: {
      stationId: z.string().min(1).describe('Print station ID'),
      payload: z.string().min(1).describe('Print payload'),
      payloadKind: z.enum(['zpl', 'pdf']).optional().describe('Payload kind (defaults to zpl)'),
      printerName: z.string().min(1).optional().describe('Target printer name'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Enqueue print job', params);
      }
      const job = await commerce.printStations.enqueueJob(params.stationId, {
        payload: params.payload,
        payloadKind: params.payloadKind,
        printerName: params.printerName,
      });
      return { success: true, message: 'Print job enqueued', job };
    },
  },
  {
    name: 'pick_up_next_print_job',
    description: 'Pick up the next queued print job for a station.',
    inputSchema: {
      stationId: z.string().min(1).describe('Print station ID'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Pick up next print job', params);
      }
      const job = await commerce.printStations.nextJob(params.stationId);
      if (!job) {
        return { success: false, error: 'No queued print job for this station' };
      }
      return { success: true, message: 'Print job picked up', job };
    },
  },
  {
    name: 'complete_print_job',
    description: 'Mark a print job printed or failed.',
    inputSchema: {
      jobId: z.string().min(1).describe('Print job ID'),
      success: z.boolean().describe('True to mark printed, false to mark failed'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Complete print job', params);
      }
      const job = await commerce.printStations.completeJob(params.jobId, params.success);
      return { success: true, message: 'Print job completed', job };
    },
  },
]);

export default printStationTools;
